core/src/test/java/io/grpc/internal/AbstractClientStreamTest.java

```diff
@@ -0,0 +1,254 @@
+  }
+
+  @Test
+  public void inboundHeadersReceived_failsOnPhaseStatus() {
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    Headers headers = new Metadata.Headers();
+    stream.inboundPhase(Phase.STATUS);
+
+    thrown.expect(IllegalStateException.class);
+
+    stream.inboundHeadersReceived(headers);
+  }
+
+  @Test
+  public void inboundHeadersReceived_succeedsOnPhaseMessage() {
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    Headers headers = new Metadata.Headers();
+    stream.inboundPhase(Phase.MESSAGE);
+
+    stream.inboundHeadersReceived(headers);
+
+    verify(mockListener).headersRead(headers);
+  }
+
+  @Test
+  public void inboundHeadersReceived_acceptsGzipEncoding() {
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    Headers headers = new Metadata.Headers();
+    headers.put(ChannelImpl.MESSAGE_ENCODING_KEY, new MessageEncoding.Gzip().getMessageEncoding());
+
+    stream.inboundHeadersReceived(headers);
+    verify(mockListener).headersRead(headers);
+  }
+
+  @Test
+  public void inboundHeadersReceived_acceptsIdentityEncoding() {
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    Headers headers = new Metadata.Headers();
+    headers.put(ChannelImpl.MESSAGE_ENCODING_KEY, MessageEncoding.NONE.getMessageEncoding());
+
+    stream.inboundHeadersReceived(headers);
+    verify(mockListener).headersRead(headers);
+  }
+
+  @Test
+  public void inboundHeadersReceived_notifiesListenerOnBadEncoding() {
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    Headers headers = new Metadata.Headers();
+    headers.put(ChannelImpl.MESSAGE_ENCODING_KEY, "bad");
+
+    stream.inboundHeadersReceived(headers);
+
+    verify(mockListener).closed(statusCaptor.capture(), isA(Metadata.class));
+    assertEquals(Code.INVALID_ARGUMENT, statusCaptor.getValue().getCode());
+  }
+
+  /**
+   * No-op base class for testing.
+   */
+  private static class BaseAbstractClientStream<T> extends AbstractClientStream<T> {
+    protected BaseAbstractClientStream(
+        WritableBufferAllocator allocator, ClientStreamListener listener) {
+      super(allocator, listener);
+    }
+
+    @Override
+    public void request(int numMessages) {}
+
+    @Override
+    protected void sendFrame(WritableBuffer frame, boolean endOfStream, boolean flush) {}
+
+    @Override
+    protected void sendCancel(Status reason) {}
+
+    @Override
+    public T id() {
+      return null;
+    }
+
+    @Override
+    protected void returnProcessedBytes(int processedBytes) {}
+  }
+
+  /**
+   * No-op base class for testing.
+   */
+  private static class BaseClientStreamListener implements ClientStreamListener {
```

Comment:
Although it doesn't matter too much, it does seem you could just use `mockListener` anywhere you use this. I do see you only use the mock when you will verify it. I figure you could just construct the AbstractClientStream as a field and get rid of the BaseAbstractClientStream construction in all but three tests.

But I'm also fine with the current organization.